src/EditorFeatures/CSharp/CommentSelection/ToggleBlockCommentCommandHandler.cs

```diff
@@ -0,0 +1,428 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CommentSelection;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Editor.Implementation.CommentSelection;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Experiments;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Commanding;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Editor.Commanding.Commands;
+using Microsoft.VisualStudio.Text.Operations;
+using Microsoft.VisualStudio.Utilities;
+using Roslyn.Utilities;
+using VSCommanding = Microsoft.VisualStudio.Commanding;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.CommentSelection
+{
+    /* TODO - Modify these once the toggle block comment handler is added.
+    [Export(typeof(VSCommanding.ICommandHandler))]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [Name(PredefinedCommandHandlerNames.CommentSelection)]*/
+    internal class ToggleBlockCommentCommandHandler :
+        AbstractCommentSelectionCommandHandler/*,
+        VSCommanding.ICommandHandler<CommentSelectionCommandArgs>*/
+    {
+        [ImportingConstructor]
+        internal ToggleBlockCommentCommandHandler(
+            ITextUndoHistoryRegistry undoHistoryRegistry,
+            IEditorOperationsFactoryService editorOperationsFactoryService) : base(undoHistoryRegistry, editorOperationsFactoryService)
+        { }
+
+        /* TODO - modify once the toggle block comment handler is added.
+        public VSCommanding.CommandState GetCommandState(CommentSelectionCommandArgs args)
+        {
+            return GetCommandState(args.SubjectBuffer);
+        }
+
+        public bool ExecuteCommand(CommentSelectionCommandArgs args, CommandExecutionContext context)
+        {
+            return ExecuteCommand(args.TextView, args.SubjectBuffer, Operation.Toggle, context);
+        }*/
+
+        public override string DisplayName => EditorFeaturesResources.Toggle_Block_Comment;
+
+        protected override string GetTitle(Operation operation)
+        {
+            return EditorFeaturesResources.Toggle_Block_Comment;
+        }
+
+        protected override string GetMessage(Operation operation)
+        {
+            return EditorFeaturesResources.Toggling_block_comment_on_selection;
+        }
+
+        protected override void SetTrackingSpans(ITextView textView, ITextBuffer buffer, List<CommentTrackingSpan> trackingSpans)
+        {
+            var spans = trackingSpans.Select(trackingSpan => trackingSpan.ToSelection(buffer));
+            textView.GetMultiSelectionBroker().SetSelectionRange(spans, spans.Last());
+        }
```

Comment:
1. what is this for?
2. how do you know the list is non-empty (for .Last to succeed)?

Reply:
1.  This is for block selection (when you have multiple carets).  This will set the multiple selections back.  I'm not 100% certain this is the correct way to do this, @olegtk could you confirm?
2.  In the abstract handler, it checks for trackingSpans.Any().  I can move it for clarity, but then I'd end up checking it in two places (this handler and the comment uncomment one)

Reply:
why do we have differing behavior for Comment/Uncomment and for Toggle here? If there is a good reason to be different, document the reason.  If there isn't a good reason to be different, we should unify.